From: Kairui Song <ryncsn@gmail.com>
Subject: mm-list_lru-split-the-lock-to-per-cgroup-scope-fix
Date: Mon, 28 Oct 2024 01:26:30 +0800

Thanks for the report. I converted list_lru_walk callback to keep the
list unlocked when LRU_RETRY and LRU_REMOVED_RETRY is returned, but
didn't notice shrink_memcg_cg in zswap.c could return LRU_STOP after
it unlocked the list.

Link: https://lkml.kernel.org/r/CAMgjq7D_OA=vYf5SnNnKXjppPFhDqsbYF--6=cOayKiadxuwrQ@mail.gmail.com
Signed-off-by: Kairui Song <ryncsn@gmail.com>
Reported-by: Usama Arif <usamaarif642@gmail.com>
Closes: https://lkml.kernel.org/r/7334b158-8567-4c66-8176-57b879ff8882@gmail.com
Cc: Chengming Zhou <zhouchengming@bytedance.com>
Cc: Johannes Weiner <hannes@cmpxchg.org>
Cc: Kairui Song <kasong@tencent.com>
Cc: Matthew Wilcox <willy@infradead.org>
Cc: Michal Hocko <mhocko@suse.com>
Cc: Muchun Song <muchun.song@linux.dev>
Cc: Qi Zheng <zhengqi.arch@bytedance.com>
Cc: Roman Gushchin <roman.gushchin@linux.dev>
Cc: Shakeel Butt <shakeel.butt@linux.dev>
Cc: Waiman Long <longman@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/list_lru.c |    9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

--- a/mm/list_lru.c~mm-list_lru-split-the-lock-to-per-cgroup-scope-fix
+++ a/mm/list_lru.c
@@ -298,9 +298,9 @@ restart:
 		ret = isolate(item, l, &l->lock, cb_arg);
 		switch (ret) {
 		/*
-		 * LRU_RETRY and LRU_REMOVED_RETRY will drop the lru lock,
-		 * the list traversal will be invalid and have to restart from
-		 * scratch.
+		 * LRU_RETRY, LRU_REMOVED_RETRY and LRU_STOP will drop the lru
+		 * lock, the list traversal will be invalid and have to restart
+		 * from scratch.
 		 */
 		case LRU_RETRY:
 			goto restart;
@@ -318,14 +318,13 @@ restart:
 		case LRU_SKIP:
 			break;
 		case LRU_STOP:
-			assert_spin_locked(&l->lock);
 			goto out;
 		default:
 			BUG();
 		}
 	}
-out:
 	unlock_list_lru(l, irq_off);
+out:
 	return isolated;
 }
 
_
